Handle wrong API formats

Currently, we assume that data is present but that is not the case when the API format is wrong. Should be handled on invoice as well as purchase order

Ex: Wrong client format
```
curl --insecure -X POST "https://localhost:8082/invoice" -H "accept: application/json" -H "Content-Type: application/json" -d '{ "recipients": ["5VVLsf+j"], "document": { "data": {"currency": "USD", "net_amount": "1501"}}}' -k
```

Expected format:
```
curl --insecure -X POST "https://localhost:8082/invoice" -H "accept: application/json" -H "Content-Type: application/json" -d '{ "collaborators": ["5VVLsf+j"], "data": { "currency":  "USD", "net_amount": "1501"}}' -k
```


Result: Node crashes instead of returning an error
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xc45689]
goroutine 596 [running]:
github.com/centrifuge/go-centrifuge/documents/invoice.(*Invoice).initInvoiceFromData(0xc42030a8c0, 0x0, 0xc42030a8c0, 0x0)
        /go/src/github.com/centrifuge/go-centrifuge/documents/invoice/model.go:177 +0x29
github.com/centrifuge/go-centrifuge/documents/invoice.(*Invoice).InitInvoiceInput(0xc42030a8c0, 0xc42041db00, 0xc42079c148, 0xc420c8e280, 0x40)
        /go/src/github.com/centrifuge/go-centrifuge/documents/invoice/model.go:160 +0x51
github.com/centrifuge/go-centrifuge/documents/invoice.service.DeriveFromCreatePayload(0x10f2ca0, 0xc4207c4fa0, 0x10f4ba0, 0xc4207be8a0, 0x10e50e0, 0x1993c10, 0x10ec8a0, 0xc4207c4980, 0xc42041db00, 0xc42079c148, ...)
        /go/src/github.com/centrifuge/go-centrifuge/documents/invoice/service.go:124 +0x6f
github.com/centrifuge/go-centrifuge/documents/invoice.(*grpcHandler).Create(0xc420283a80, 0x10f17e0, 0xc4203e8a50, 0xc42041db00, 0xc420283a80, 0xc4203e89f0, 0xd91940)
        /go/src/github.com/centrifuge/go-centrifuge/documents/invoice/handler.go:43 +0x217
github.com/centrifuge/go-centrifuge/protobufs/gen/go/invoice._DocumentService_Create_Handler(0xdcd3e0, 0xc420283a80, 0x10f17e0, 0xc4203e8a50, 0xc4202be2a0, 0x0, 0x0, 0x0, 0xc42049b560, 0x0)
        /go/src/github.com/centrifuge/go-centrifuge/protobufs/gen/go/invoice/service.pb.go:643 +0x241
github.com/centrifuge/go-centrifuge/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc420c64600, 0x10f6ee0, 0xc4203e65a0, 0xc42031da00, 0xc42030d8f0, 0x1873600, 0x0, 0x0, 0x0)
        /go/src/github.com/centrifuge/go-centrifuge/vendor/google.golang.org/grpc/server.go:982 +0x4f9
github.com/centrifuge/go-centrifuge/vendor/google.golang.org/grpc.(*Server).handleStream(0xc420c64600, 0x10f6ee0, 0xc4203e65a0, 0xc42031da00, 0x0)
        /go/src/github.com/centrifuge/go-centrifuge/vendor/google.golang.org/grpc/server.go:1208 +0x1318
github.com/centrifuge/go-centrifuge/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc4204c0040, 0xc420c64600, 0x10f6ee0, 0xc4203e65a0, 0xc42031da00)
        /go/src/github.com/centrifuge/go-centrifuge/vendor/google.golang.org/grpc/server.go:686 +0x9f
created by github.com/centrifuge/go-centrifuge/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
        /go/src/github.com/centrifuge/go-centrifuge/vendor/google.golang.org/grpc/server.go:684 +0xa1
```

Acceptance criteria:
 - Proper nil checks for full round trip i.e handlers -> service -> handler for all APIs
 - Integrartion tests in go-centrifuge
 - functional tests that sends a wrong format and error is received instead of node crash

--- a/documents/invoice/handler.go
+++ b/documents/invoice/handler.go
@@ -89,21 +89,25 @@ func (h *grpcHandler) GetVersion(ctx context.Context, getVersionRequest *clienti
 		apiLog.Error(err)
 		return nil, centerrors.Wrap(err, "identifier is invalid")
 	}
+
 	version, err := hexutil.Decode(getVersionRequest.Version)
 	if err != nil {
 		apiLog.Error(err)
 		return nil, centerrors.Wrap(err, "version is invalid")
 	}
+
 	model, err := h.service.GetVersion(identifier, version)
 	if err != nil {
 		apiLog.Error(err)
 		return nil, centerrors.Wrap(err, "document not found")
 	}
+
 	resp, err := h.service.DeriveInvoiceResponse(model)
 	if err != nil {
 		apiLog.Error(err)
 		return nil, err
 	}
+
 	return resp, nil
 }
 
@@ -115,15 +119,18 @@ func (h *grpcHandler) Get(ctx context.Context, getRequest *clientinvoicepb.GetRe
 		apiLog.Error(err)
 		return nil, centerrors.Wrap(err, "identifier is an invalid hex string")
 	}
+
 	model, err := h.service.GetCurrentVersion(identifier)
 	if err != nil {
 		apiLog.Error(err)
 		return nil, centerrors.Wrap(err, "document not found")
 	}
+
 	resp, err := h.service.DeriveInvoiceResponse(model)
 	if err != nil {
 		apiLog.Error(err)
 		return nil, err
 	}
+
 	return resp, nil
 }

--- a/documents/invoice/service.go
+++ b/documents/invoice/service.go
@@ -115,13 +115,13 @@ func (s service) DeriveFromCoreDocument(cd *coredocumentpb.CoreDocument) (docume
 }
 
 // UnpackFromCreatePayload initializes the model with parameters provided from the rest-api call
-func (s service) DeriveFromCreatePayload(invoiceInput *clientinvoicepb.InvoiceCreatePayload, contextHeader *documents.ContextHeader) (documents.Model, error) {
-	if invoiceInput == nil {
+func (s service) DeriveFromCreatePayload(payload *clientinvoicepb.InvoiceCreatePayload, contextHeader *documents.ContextHeader) (documents.Model, error) {
+	if payload == nil || payload.Data == nil {
 		return nil, centerrors.New(code.DocumentInvalid, "input is nil")
 	}
 
 	invoiceModel := new(Invoice)
-	err := invoiceModel.InitInvoiceInput(invoiceInput, contextHeader)
+	err := invoiceModel.InitInvoiceInput(payload, contextHeader)
 	if err != nil {
 		return nil, centerrors.New(code.DocumentInvalid, err.Error())
 	}
@@ -283,13 +283,12 @@ func (s service) DeriveInvoiceData(doc documents.Model) (*clientinvoicepb.Invoic
 		return nil, centerrors.New(code.DocumentInvalid, "document of invalid type")
 	}
 
-	data := inv.getClientData()
-	return data, nil
+	return inv.getClientData(), nil
 }
 
 // DeriveFromUpdatePayload returns a new version of the old invoice identified by identifier in payload
 func (s service) DeriveFromUpdatePayload(payload *clientinvoicepb.InvoiceUpdatePayload, contextHeader *documents.ContextHeader) (documents.Model, error) {
-	if payload == nil {
+	if payload == nil || payload.Data == nil {
 		return nil, centerrors.New(code.DocumentInvalid, "invalid payload")
 	}
 

--- a/documents/invoice/service_test.go
+++ b/documents/invoice/service_test.go
@@ -98,6 +98,10 @@ func TestService_DeriveFromPayload(t *testing.T) {
 	_, err = invSrv.DeriveFromCreatePayload(nil, nil)
 	assert.Error(t, err, "DeriveWithInvoiceInput should produce an error if invoiceInput equals nil")
 
+	// fail due to nil payload data
+	_, err = invSrv.DeriveFromCreatePayload(&clientinvoicepb.InvoiceCreatePayload{}, nil)
+	assert.Error(t, err, "DeriveWithInvoiceInput should produce an error if invoiceInput equals nil")
+
 	contextHeader, err := documents.NewContextHeader()
 	assert.Nil(t, err)
 	model, err = invSrv.DeriveFromCreatePayload(payload, contextHeader)
@@ -474,6 +478,12 @@ func TestService_DeriveFromUpdatePayload(t *testing.T) {
 	invSrv := getServiceWithMockedLayers()
 	// nil payload
 	doc, err := invSrv.DeriveFromUpdatePayload(nil, nil)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "invalid payload")
+	assert.Nil(t, doc)
+
+	// nil payload data
+	doc, err = invSrv.DeriveFromUpdatePayload(&clientinvoicepb.InvoiceUpdatePayload{}, nil)
 	assert.Error(t, err)
 	assert.Contains(t, err.Error(), "invalid payload")
 	assert.Nil(t, doc)

--- a/documents/invoice/service_test.go
+++ b/documents/invoice/service_test.go
@@ -491,7 +491,7 @@ func TestService_DeriveFromUpdatePayload(t *testing.T) {
 	// messed up identifier
 	contextHeader, err := documents.NewContextHeader()
 	assert.Nil(t, err)
-	payload := &clientinvoicepb.InvoiceUpdatePayload{Identifier: "some identifier"}
+	payload := &clientinvoicepb.InvoiceUpdatePayload{Identifier: "some identifier", Data: &clientinvoicepb.InvoiceData{}}
 	doc, err = invSrv.DeriveFromUpdatePayload(payload, contextHeader)
 	assert.Error(t, err)
 	assert.Contains(t, err.Error(), "failed to decode identifier")

--- a/documents/purchaseorder/handler.go
+++ b/documents/purchaseorder/handler.go
@@ -90,21 +90,25 @@ func (h grpcHandler) GetVersion(ctx context.Context, req *clientpurchaseorderpb.
 		apiLog.Error(err)
 		return nil, centerrors.Wrap(err, "identifier is invalid")
 	}
+
 	version, err := hexutil.Decode(req.Version)
 	if err != nil {
 		apiLog.Error(err)
 		return nil, centerrors.Wrap(err, "version is invalid")
 	}
+
 	model, err := h.service.GetVersion(identifier, version)
 	if err != nil {
 		apiLog.Error(err)
 		return nil, centerrors.Wrap(err, "document not found")
 	}
+
 	resp, err := h.service.DerivePurchaseOrderResponse(model)
 	if err != nil {
 		apiLog.Error(err)
 		return nil, err
 	}
+
 	return resp, nil
 }
 
@@ -116,15 +120,18 @@ func (h grpcHandler) Get(ctx context.Context, getRequest *clientpurchaseorderpb.
 		apiLog.Error(err)
 		return nil, centerrors.Wrap(err, "identifier is an invalid hex string")
 	}
+
 	model, err := h.service.GetCurrentVersion(identifier)
 	if err != nil {
 		apiLog.Error(err)
 		return nil, centerrors.Wrap(err, "document not found")
 	}
+
 	resp, err := h.service.DerivePurchaseOrderResponse(model)
 	if err != nil {
 		apiLog.Error(err)
 		return nil, err
 	}
+
 	return resp, nil
 }

--- a/documents/purchaseorder/service.go
+++ b/documents/purchaseorder/service.go
@@ -143,7 +143,7 @@ func (s service) Update(ctx context.Context, po documents.Model) (documents.Mode
 
 // DeriveFromCreatePayload derives purchase order from create payload
 func (s service) DeriveFromCreatePayload(payload *clientpopb.PurchaseOrderCreatePayload, ctxH *documents.ContextHeader) (documents.Model, error) {
-	if payload == nil {
+	if payload == nil || payload.Data == nil {
 		return nil, centerrors.New(code.DocumentInvalid, "input is nil")
 	}
 
@@ -158,7 +158,7 @@ func (s service) DeriveFromCreatePayload(payload *clientpopb.PurchaseOrderCreate
 
 // DeriveFromUpdatePayload derives purchase order from update payload
 func (s service) DeriveFromUpdatePayload(payload *clientpopb.PurchaseOrderUpdatePayload, ctxH *documents.ContextHeader) (documents.Model, error) {
-	if payload == nil {
+	if payload == nil || payload.Data == nil {
 		return nil, centerrors.New(code.DocumentInvalid, "invalid payload")
 	}
 

--- a/documents/purchaseorder/service_test.go
+++ b/documents/purchaseorder/service_test.go
@@ -142,7 +142,7 @@ func TestService_DeriveFromUpdatePayload(t *testing.T) {
 	// messed up identifier
 	contextHeader, err := documents.NewContextHeader()
 	assert.Nil(t, err)
-	payload := &clientpurchaseorderpb.PurchaseOrderUpdatePayload{Identifier: "some identifier"}
+	payload := &clientpurchaseorderpb.PurchaseOrderUpdatePayload{Identifier: "some identifier", Data: &clientpurchaseorderpb.PurchaseOrderData{}}
 	doc, err = poSrv.DeriveFromUpdatePayload(payload, contextHeader)
 	assert.Error(t, err)
 	assert.Contains(t, err.Error(), "failed to decode identifier")

--- a/documents/purchaseorder/service_test.go
+++ b/documents/purchaseorder/service_test.go
@@ -133,6 +133,12 @@ func TestService_DeriveFromUpdatePayload(t *testing.T) {
 	assert.Contains(t, err.Error(), "invalid payload")
 	assert.Nil(t, doc)
 
+	// nil payload data
+	doc, err = poSrv.DeriveFromUpdatePayload(&clientpurchaseorderpb.PurchaseOrderUpdatePayload{}, nil)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "invalid payload")
+	assert.Nil(t, doc)
+
 	// messed up identifier
 	contextHeader, err := documents.NewContextHeader()
 	assert.Nil(t, err)
@@ -205,6 +211,12 @@ func TestService_DeriveFromCreatePayload(t *testing.T) {
 
 	// nil payload
 	m, err := poSrv.DeriveFromCreatePayload(nil, ctxh)
+	assert.Nil(t, m)
+	assert.Error(t, err)
+	assert.Contains(t, err.Error(), "input is nil")
+
+	// nil data payload
+	m, err = poSrv.DeriveFromCreatePayload(&clientpurchaseorderpb.PurchaseOrderCreatePayload{}, ctxh)
 	assert.Nil(t, m)
 	assert.Error(t, err)
 	assert.Contains(t, err.Error(), "input is nil")

